### PR TITLE
Add ddpm model

### LINKED ISSUE #55: Integrate Maria's Model (DDPM) into main branch
### Problem / Goal

We need to integrate the DDPM model from the `mc-models` branch into the `main` branch.
This ensures the new model is available in the main development workflow and aligns with our upcoming integration milestone.

### What does good look like?

- [ ] DDPM model code fully merged into `main` without breaking existing functionality.
- [ ] Hydra config updates from `mc-models` are included and functional.
- [ ] All tests pass successfully.
- [ ] Sanity check confirms no significant change in results unless intended.

### Related context

`mc-models` branch with DDPM implementation: https://github.com/alan-turing-institute/ice-station-zebra/tree/mc-models
`main` branch: https://github.com/alan-turing-institute/ice-station-zebra

### PR DESCRIPTION
**Pull request by marianovitasari20**
Closes #55

### CODE REVIEW

File: ice_station_zebra/models/diffusion/gaussian_diffusion.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            timesteps (int): Total number of timesteps.
+            s (float): Small offset to prevent singularities near 0.
+                      Controls how fast the signal decays: a smaller s results in faster
+                      corruption early on; a larger s gives a smoother, slower decay.
+
+        Returns:
+            torch.Tensor: Beta values of shape (timesteps,).
+        """
+        steps = timesteps + 1
+        x = torch.linspace(0, timesteps, steps)
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply ice_station_zebra/models/diffusion/gaussian_diffusion.py:86]: Or you could even do the scaling from 0-1 here too and do `x = torch.arange(0, timesteps + 1) / timesteps`

File: ice_station_zebra/models/diffusion/gaussian_diffusion.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        alphas_cumprod = torch.cos(((x / timesteps) + s) / (1 + s) * math.pi * 0.5) ** 2
```
Comment (jemrobinson) [inline ice_station_zebra/models/diffusion/gaussian_diffusion.py:87]:
```suggestion
        steps = timesteps + 1
        t = torch.arange(0, timesteps + 1) / timesteps
        alphas_cumprod = torch.cos((t + s) / (1 + s) * math.pi * 0.5) ** 2
```

File: ice_station_zebra/models/diffusion/gaussian_diffusion.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        alphas_cumprod = alphas_cumprod / alphas_cumprod[0]
+        betas = 1 - (alphas_cumprod[1:] / alphas_cumprod[:-1])
+
+        return torch.clip(betas, 0, 0.999)
+
+    def q_sample(
```
Comment (jemrobinson) [inline ice_station_zebra/models/diffusion/gaussian_diffusion.py:93]: Is this function used anywhere?

File: ice_station_zebra/models/diffusion/unet_diffusion.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+            noise (torch.Tensor): Noisy forecast tensor of shape [B, H, W, n_classes, n_forecast_days].
+            t (torch.Tensor): Diffusion timestep tensor of shape [B].
+            conditioning (torch.Tensor): Conditioning input tensor of shape [B, H, W, input_channels].
+
+        Returns:
+            torch.Tensor: Predicted denoised forecast of shape [B, H, W, n_classes, n_forecast_days].
+
+        """
+        noise = 2.0 * noise - 1.0
+        conditioning = 2.0 * conditioning - 1.0
```
Comment (jemrobinson) [inline ice_station_zebra/models/diffusion/unet_diffusion.py:123]: I assume this is to convert a range 0 to 1 to -1 to 1? Are we sure that both `noise` and `conditioning` are in the range 0-1? We should probably add this assumption to the docstring too.

File: ice_station_zebra/models/processors/ddpm.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+        Args:
+            n_latent_channels: Num latent channels used in UNetDiffusion.
+            timesteps: Num diffusion timesteps (default=1000).
+
+        """
+        super().__init__()
+        self.model = UNetDiffusion(n_latent_channels, timesteps)
+        self.timesteps = timesteps
+        self.ema = ExponentialMovingAverage(self.model.parameters(), decay=0.995)
```
Comment (jemrobinson) [inline ice_station_zebra/models/processors/ddpm.py:27]: I think we don't need EMA here any more.

File: ice_station_zebra/models/processors/ddpm.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            torch.Tensor: Denoised output of shape [B, H, W, C].
+
+        """
+        # Start from pure noise
+        y = torch.rand_like(x)
+
+        dim_threshold = 3
+
+        # Use EMA weights for sampling
+        with self.ema.average_parameters():
```
Comment (jemrobinson) [inline ice_station_zebra/models/processors/ddpm.py:71]: I think this should be handled by the EMACallback.

File: ice_station_zebra/models/diffusion/unet_diffusion.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+        # Time embedding
+        t = self._timestep_embedding(t)
+        t = self.time_embed(t)
+
+        # Concatenate with conditional input
+        noise = torch.cat([noise, conditioning], dim=-1)  # [b,h,w,(d*c)+input_channels]
+
+        # Convert to channel-first format
+        noise = torch.movedim(noise, -1, 1)  # [b,channels,h,w]
```
Comment (jemrobinson) [inline ice_station_zebra/models/diffusion/unet_diffusion.py:133]: We should change to requiring the inputs in `[B, C, H, W]` format and drop this dimension switch.

File: ice_station_zebra/models/diffusion/unet_diffusion.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+        up9 = self.up9(up8)
+        up9 = torch.cat([bn1, up9], dim=1)
+        up9 = self._add_time_embedding(up9, t)
+        up9 = self.up9b(up9)
+
+        # Final output
+        output = self.final_layer(up9)  # [b, c_out, h, w]
+
+        return torch.movedim(output, 1, -1)  # [b, h, w, c_out]
```
Comment (jemrobinson) [inline ice_station_zebra/models/diffusion/unet_diffusion.py:172]: We should change to providing the outputs in `[B, C, H, W]` format and drop this dimension switch.

File: ice_station_zebra/models/processors/ddpm.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        """Transformation summary.
+
+        Args:
+            x: TensorNCHW with (batch_size, latent_channels, latent_height, latent_width)
+
+        Returns:
+            TensorNCHW with (batch_size, latent_channels, latent_height, latent_width)
+
+        """
+        x_bhwc = x.movedim(1, -1)
```
Comment (jemrobinson) [inline ice_station_zebra/models/processors/ddpm.py:40]: We should keep this in `[B, C, H, W]` format, updating `UNetDiffusion` accordingly.

File: pyproject.toml [hunk 10]
```diff
@@ -23,6 +23,7 @@ dependencies = [
     "netcdf4>=1.7.0",
     "pillow>=11.3.0",
     "pydantic>=2.11.7",
+    "torch-ema>=0.3",
```
Comment (jemrobinson) [inline pyproject.toml:26]: I don't think we need this.

File: ice_station_zebra/models/diffusion/gaussian_diffusion.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    def p_sample(
+        self, x: torch.Tensor, t: torch.Tensor, pred_v: torch.Tensor
+    ) -> torch.Tensor:
+        """Perform a single reverse diffusion (denoising) step using the v-prediction parameterization.
+
+        This method implements the reverse process using v-prediction rather than epsilon-prediction,
+        where the model predicts velocity v_t instead of noise epsilon.
+
+        Args:
+            x (torch.Tensor): Current noisy sample x_t at timestep t (shape: [B,H,W,n_classes*n_forecast_days]).
```
Comment (jemrobinson) [inline ice_station_zebra/models/diffusion/gaussian_diffusion.py:99]: We should update this to use `[B, C, H, W]` format.

**Review by jemrobinson — CHANGES_REQUESTED** [review]
We should use `B, C, H, W` format throughout the diffusion process. Happy to help with implementing this if it's not clear how to do it.

We also don't need EMA to be implemented inside the model, as we're running EMA in a lightning callback.

File: ice_station_zebra/models/common/timeembed.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        """
+        super().__init__()
+
+        activation_layer = ACTIVATION_FROM_NAME[activation]
+
+        self.model = nn.Sequential(
+            nn.Linear(dim, dim * 4),
+            activation_layer(inplace=True),
+            nn.Linear(dim * 4, dim),
+        )
```
[the remark this answers is not in the document]
Comment (marianovitasari20) [reply ice_station_zebra/models/common/timeembed.py:29]: Hi @jemrobinson,
I was thinking of pushing this first and then adjusting some blocks later on separate branches, so that I have smaller modules that are more manageable and easier to test. I’ll start by changing the common block, might update the encoder decoder for diffusion, and then adjust based on the input (either NCHW or NTCHW). What do you think?

[the remark this answers is not in the document]
Comment (jemrobinson) [reply ice_station_zebra/models/common/timeembed.py:29]: That's absolutely fine - in fact it's better to break up a PR into smaller ones when possible. Can you open an issue for it and link that here?

File: ice_station_zebra/models/diffusion/unet_diffusion.py [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+        """
+        super().__init__()
+
+        self.filter_size = filter_size
+        self.start_out_channels = start_out_channels
+        self.timesteps = timesteps
+
+        # Time embedding
+        self.time_embed_dim = 256
```
[the remark this answers is not in the document]
Comment (marianovitasari20) [reply ice_station_zebra/models/diffusion/unet_diffusion.py:54]: I see your point. I tend to keep it as a `self` member so that it’s visible as a property of the model and can be accessed later if needed.

[the remark this answers is not in the document]
Comment (jemrobinson) [reply ice_station_zebra/models/diffusion/unet_diffusion.py:54]: Maybe it's better to make it an argument to the constructor (with a default value of 256), so we can change it in the config if needed? I think this might also mean that it gets logged to W&B (but I haven't checked that).

File: ice_station_zebra/models/common/timeembed.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (marianovitasari20) [reply ice_station_zebra/models/common/timeembed.py:29]: This MLP isn’t making the time embedding from scratch, it’s actually the second step in the pipeline.

If you check `UNetDiffusion`, the flow is:
1. `_timestep_embedding()` turns the integer timesteps into sinusoidal embeddings
2. `TimeEmbed` (this class) takes those and learns a more useful representation
Basically, the MLP is just a projection layer that lets the model adapt the fixed sinusoidal stuff into something better for our diffusion task. The dim→4\*dim→dim expansion is a common trick to give it more flexibility.

File: ice_station_zebra/models/diffusion/gaussian_diffusion.py (hunk 11 above)
Comment (jemrobinson) [reply ice_station_zebra/models/diffusion/gaussian_diffusion.py:99]: @marianovitasari20: can you open an issue for this, link to it here and then resolve the conversation?

File: ice_station_zebra/models/diffusion/unet_diffusion.py (hunk 4 above)
Comment (jemrobinson) [reply ice_station_zebra/models/diffusion/unet_diffusion.py:123]: @marianovitasari20: if you open an issue for adding tests, this could be part of that.

File: ice_station_zebra/models/diffusion/gaussian_diffusion.py (hunk 11 above)
Comment (marianovitasari20) [reply ice_station_zebra/models/diffusion/gaussian_diffusion.py:99]: https://github.com/alan-turing-institute/ice-station-zebra/issues/82

File: ice_station_zebra/models/processors/ddpm.py (hunk 9 above)
Comment (marianovitasari20) [reply ice_station_zebra/models/processors/ddpm.py:40]: https://github.com/alan-turing-institute/ice-station-zebra/issues/82

File: ice_station_zebra/models/diffusion/unet_diffusion.py (hunk 8 above)
Comment (marianovitasari20) [reply ice_station_zebra/models/diffusion/unet_diffusion.py:172]: https://github.com/alan-turing-institute/ice-station-zebra/issues/82

File: ice_station_zebra/models/diffusion/gaussian_diffusion.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+            a (torch.Tensor): 1D tensor containing precomputed values (e.g., alpha or beta schedule).
+            t (torch.Tensor): Timesteps for each sample in the batch (shape: [B]).
+            x_shape (Tuple[int]): Target shape for broadcasting (same as input sample x).
+
+        Returns:
+            torch.Tensor: Extracted and reshaped values for each timestep in the batch.
+
+        """
+        a = a.to(t.device)
+        out = a[t]  # (batch_size,) # Reshape for broadcasting: [batch_size, 1, 1, 1, 1]
```
[the remark this answers is not in the document]
Comment (marianovitasari20) [reply ice_station_zebra/models/diffusion/gaussian_diffusion.py:144]: https://github.com/alan-turing-institute/ice-station-zebra/issues/83

File: ice_station_zebra/models/common/timeembed.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (marianovitasari20) [reply ice_station_zebra/models/common/timeembed.py:29]: https://github.com/alan-turing-institute/ice-station-zebra/issues/84

File: pyproject.toml (hunk 10 above)
Comment (marianovitasari20) [reply pyproject.toml:26]: https://github.com/alan-turing-institute/ice-station-zebra/issues/84

File: ice_station_zebra/models/common/timeembed.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (marianovitasari20) [reply ice_station_zebra/models/common/timeembed.py:34]: > OK, that makes sense. Is there a better way to explain this in the docstring? Currently it says "Initialize the time embedding module", but it might be good to add something that explains why we're projecting the sinusoidal timesteps though some fully-connected layers.

Thanks @jemrobinson! I’ve updated the docstring.

**Review by jemrobinson — APPROVED** [review]
LGTM

File: ice_station_zebra/models/diffusion/unet_diffusion.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (jemrobinson) [reply ice_station_zebra/models/diffusion/unet_diffusion.py]: Done in 194525e

File: ice_station_zebra/models/diffusion/unet_diffusion.py (hunk 4 above)
Comment (jemrobinson) [reply ice_station_zebra/models/diffusion/unet_diffusion.py:124]: #83

File: ice_station_zebra/models/processors/ddpm.py (hunk 5 above)
Comment (jemrobinson) [reply ice_station_zebra/models/processors/ddpm.py:27]: #84

File: ice_station_zebra/models/processors/ddpm.py (hunk 6 above)
Comment (jemrobinson) [reply ice_station_zebra/models/processors/ddpm.py:71]: #84
